[Jerome Wong Jia Jin] iP

jeromewjj opened the pull request:
# Hello!!! This is Duke.

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

Duke frees your mind of having to remember things you need to do. It's,

* text-based
* easy to learn
* ~~FAST~~ _SUPER_ FAST to use

All you need to do is,

1. download it from here.
1. double-click it.
1. add your tasks.
1. let it manage your tasks for you :stuck_out_tongue:

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---
If you are Java programmer, you can use it to practice Java too. Here's the `main` method:

```c
    public static void main(String[] args) {
        new Duke("data.txt").run();
    }
```

Commit:
Add Gradle support

jeromewjj committed:
Completed Level-1. Greet, Echo, Exit

jeromewjj committed:
Added add, list function

jeromewjj committed:
Added Task.java

jeromewjj committed:
Complete Level 3. Mark as done

jeromewjj committed:
Added Todo, Event, Deadline class

jeromewjj committed:
Complete Level 4. ToDos, Events, Deadlines

jeromewjj committed:
Updated Duke and Task

jeromewjj committed:
Added input and expected input

jeromewjj committed:
Added handle error

jeromewjj committed:
Added delete

jeromewjj committed:
Added send message

jeromewjj committed:
Added javadocs comments

jeromewjj committed:
added another exception

jeromewjj committed:
Add Storage class

Storage allows task to be stored in hard drive.

The toFileString method added in Task, Event, Deadline and Todo to convert task into string to store in file.

jeromewjj committed:
Update date of Dateline class

The new date format allows for future new methods like comparing dates.

Change date from String to LocalDate.

jeromewjj committed:
Add data.txt into .gitignore

jeromewjj committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Add Storage class

jeromewjj committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Update date of Dateline class

jeromewjj committed:
Reorganise Classes to follow OOP

Command, Ui, TaskList classes are added to make Duke follow the OOP.

jeromewjj committed:
Divide classes into packages

Duke, command, parser, exception, storage, task, ui package added.

jeromewjj committed:
Add junit test for Parser and Deadline class

Test Parser method throw appropriate error.
Test Deadline toString and toFileString method.

jeromewjj committed:
Package the App as a JAR file

jeromewjj committed:
Add JavaDocs

jeromewjj committed:
Update Task name

jeromewjj committed:
Tweak the code to comply with a coding standard

jeromewjj committed:
Add find command

Find command allows user to search task using a keyword.

FindCommand class is added.

jeromewjj committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Tweak the code to comply with a coding standard

jeromewjj committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add find command

jeromewjj committed:
Add JavaDocs to FindCommand class

jeromewjj commented:
# Hello!!! This is Duke.

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

Duke frees your mind of having to remember things you need to do. It's,

* text-based
* easy to learn
* ~~FAST~~ _SUPER_ FAST to use

All you need to do is,

1. download it from here.
1. double-click it.
1. add your tasks.
1. let it manage your tasks for you :stuck_out_tongue:

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---
If you are Java programmer, you can use it to practice Java too. Here's the `main` method:

```c
    public static void main(String[] args) {
        new Duke("data.txt").run();
    }
```

jeromewjj committed:
Merge branch 'add-gradle-support' of https://github.com/jeromewjj/ip

* 'add-gradle-support' of https://github.com/jeromewjj/ip:
  Add Gradle support

jeromewjj committed:
Add checkstyle.xml

jeromewjj committed:
Move test file

jeromewjj committed:
Add test in DeadlineTest

jeromewjj committed:
Update duke to follow CheckStyle guide

jeromewjj committed:
Add GUI to Duke

jeromewjj committed:
Update GUI for duke

jeromewjj committed:
Update JavaDocs

GWphua commented on src/main/java/duke/DialogBox.java:
```diff
@@ -0,0 +1,75 @@
+    /**
+     * Flips the dialog box such that the ImageView is on the left and text on the right.
+     */
+    private void flip() {
```
Perhaps try something more specific? Like `flipDialogBox` instead?

GWphua commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,75 @@
+     * @throws DukeException if command created has error.
+     */
+    public static Command parse(String fullCommand) throws DukeException {
+        String command = fullCommand.split(" ")[0];
```
Interesting use of string manipulation!! 👍

GWphua commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,129 @@
+     *
+     * @return Exit message when user exit the programs.
+     */
+    public String showExit() {
```
Maybe `exitMessage` would be much more specific to the intentions of this method?

GWphua commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,129 @@
+     * @param taskList The list of tasks.
+     * @return Task added to the list.
+     */
+    public String showAddTask(Task task, TaskList taskList) {
```
Maybe an alternative to `show` would be better, as this is not printing (and thus showing) anything.

GWphua reviewed (commented):
LGTM, I had learnt a lot by reading your code. Thanks for writing such neat code 💯 👍

pranav-ganesh commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,75 @@
+        String command = fullCommand.split(" ")[0];
+        Command newCommand;
+
+        switch(command) {
```
I like the fact that you used a switch statement rather than if else! 😄

pranav-ganesh commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,77 @@
+    public boolean isKeyWordPresent(String keyWord) {
+        int intIndex = description.indexOf(keyWord);
+
+        if (intIndex == -1) {
```
Perhaps replace lines 70-75 with _return intIndex != -1;_ ?

pranav-ganesh commented on src/main/java/duke/command/DoneCommand.java:
```diff
@@ -0,0 +1,50 @@
+     * @return Boolean whether command is an exit command.
+     */
+    @Override
+    public boolean isExit() {
```
Maybe you can simply create a private variable called isExit and initialize it in the constructor to true/false depending on the subclass rather than having a method on every single subclass of the command class? I believe it will save many lines of code.

pranav-ganesh commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,136 @@
+     *
+     * @return Number of tasks stored in the task list.
+     */
+    public int totalTask() {
```
Perhaps a better idea to use a verb (action word) in the method name to improve clarity? How about something like getTotalTask instead of totalTask? This is mentioned in the coding standard guideline, please do take a look at it!

pranav-ganesh commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,129 @@
+     *
+     * @return Lost message that tells user that Duke can't understand the input.
+     */
+    public String showLost() {
```
Maybe simply return the string instead of creating an unnecessary variable? Saves a line of code 😄

pranav-ganesh commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,27 @@
+     */
+    public abstract String execute(TaskList tasks, Ui ui, Storage storage) throws DukeException;
+
+    /**
```
Not 100% sure about this: since this is an abstract method, it doesn't have a return value right? Are you sure @return should be included in the javadoc?

pranav-ganesh reviewed (commented):
Overall you have done a great job and the code quality is good! Just a few things here and there to take note of to further improve the code quality! Good job 👍

jeromewjj committed:
Update main class name

jeromewjj committed:
Add assertions to define assumptions in methods

Assertion is added to identify potential bugs in the methods.

Let's add assertion in all Command Classes to verify that tasks, ui and storage are not null and Task class to verify that task number are not less than 1.

jeromewjj committed:
Improve code quality in various classes

Codes were updated to improve the readability of the code.

Split various long methods into smaller method.
Change if-else statement in AddCommand constructor into case statement.

Using case statement in AddCommand constructor make codes easier to read and allow future new task class to be added easily.

jeromewjj committed:
Merge pull request #1 from jeromewjj/branch-A-Assertions

Add assertions to define assumptions in methods

jeromewjj committed:
Merge branch 'master' into branch-A-CodeQuality

jeromewjj committed:
Merge pull request #2 from jeromewjj/branch-A-CodeQuality

Improve code quality in various classes

jeromewjj committed:
Improve code quality in Ui class

Improving code quality allow code to be more readable.

Split string into multiple string variables to increase readability.

jeromewjj committed:
Implement Tag Command

jeromewjj committed:
Update code to follow checkstyle

jeromewjj committed:
Polish GUI

jeromewjj committed:
Update method name

jeromewjj committed:
Add Ui.png

jeromewjj committed:
Update Readme.md

jeromewjj committed:
Update code to follow check style guide

jeromewjj committed:
Update Java docs to follow checkstyle
